[1단계 - 출석 구현] 플린트(이주환) 미션 제출합니다.

jhan0121 opened the pull request:
안녕하세요 이프! 우아한테크코스 7기 백엔드 플린트입니다!
아래는 이번 과제를 수행하면서 작성한 체크리스트입니다!

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3945


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
Indent를 지키지 못한 메서드가 있었고, stream을 사용할 때, 메서드마다 개행을 해주는 컨벤션을 지키지 못했습니다. 
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
`컨트롤러에서 도메인에 객체를 생성한 다음 객체를 메서드에 넘기는 방법`과 `사용할 도메인 내에서 필요한 데이터를 넘기고 메서드 내에서 필요한 객체를 생성하는 방법` 중 어떤 방법이 더 좋을지 고민이 많았습니다. 개인적으로는 tdd를 진행할 때, 객체를 생성해서 메서드에 넘겨주는 방법이 테스트 코드를 작성할 떄 더 편한 느낌이었습니다. 해당 요소에 대해서 이프의 생각이 궁금합니다!

예시:

```java
Class Controller {
    public void run() {
        Item item = new Item("사과", 1);
        Shop.addItem(item);
    }
}
```

```java
Class Controller {
    public void run() {
        String name = "사과";
        int count = 1;
        Shop.addItem(name, count);
    }
}
```

jhan0121 committed:
docs: 요구사항 및 기능사항 추가

jhan0121 committed:
feat(Attend): 출석 관련 데이터를 관리하는 객체 추가

jhan0121 committed:
feat: 출석 가능 일자 확인 기능 추가

jhan0121 committed:
feat: 출석 추가 기능 구현

jhan0121 committed:
docs: CSV 고려 사항 추가

jhan0121 committed:
feat: 쉬는날 출석 시 예외 구현 추가

jhan0121 committed:
feat: 같은 날짜에 출석이 존재하는지 확인하는 기능 추가

jhan0121 committed:
feat: 출석 수정 기능 추가

jhan0121 committed:
feat: 대상 닉네임의 출석 수정 기능 추가

jhan0121 committed:
feat: 운영 시간 외 시간을 입력했을 경우 예외 처리 추가

jhan0121 committed:
fix: attend를 수정하면서 실패한 테스트 수정

jhan0121 committed:
test: findByDay 누락 테스트 추가

jhan0121 committed:
feat: 현재 날짜 이전까지 출석을 조회하는 기능 추가

jhan0121 committed:
feat: 닉네임 대상의 어제까지의 출석 정보를 반환하는 기능 추가

jhan0121 committed:
docs: csv 타입의 출석 데이터 파일 추가

jhan0121 committed:
feat: csv 타입의 파일을 읽어서 AttendanceBook 을 반환하는 기능

jhan0121 committed:
feat: Attend 와 LateTime, AbsenceTime 을 기반으로 출결을 판정하는 기능

jhan0121 committed:
feat: 이름과 날짜를 기반으로 출석과 출결 현황을 반환하는 기능 추가

jhan0121 committed:
feat: 출석, 지각, 결석 횟수를 바탕으로 제적 위험자를 파악하는 기능 추가

jhan0121 committed:
fix: 멤버 변수 타입(int -> long) 수정

jhan0121 committed:
feat: 출석 기록에서 출석, 지각, 결석 횟수를 파악하는 기능

jhan0121 committed:
refector: AttendanceResults를 일급 컬렉션으로 사용하도록 수정

jhan0121 committed:
feat: 출석부에서 제적 위험자 조회 기능 추가

jhan0121 committed:
docs: 누락한 기능 구현 사항 수정

jhan0121 committed:
feat: 출석부에 등록되지 않은 크루 닉네임을 사용하면 예외 발생 기능 추가

jhan0121 committed:
refactor: 역할에 맞는 패키지 이동

jhan0121 committed:
feat: InputView 추가

jhan0121 committed:
feat: OutputView 추가

jhan0121 committed:
feat: Controller 구현 및 기능 통합

sinb57 commented on src/main/java/util/DateUtil.java:
```diff
@@ -0,0 +1,70 @@
+
+public class DateUtil {
+
+    public static LocalDateTime parsetime(String day, String time) {
```
DateUtilTest 에서만 사용되는 메서드네요! 필요한가요??

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,106 @@
+
+            } catch (IllegalArgumentException e) {
+                System.out.println(e.getMessage());
+            }
```
> 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.

요구사항이 지켜지지 않았네요😢 수정 부탁드려요

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,106 @@
+                if (command.equals("2")) {
+                    String nickName = inputView.inputEditNickName();
+                    String date = inputView.inputDate();
+                    String time = inputView.inputEditTime();
```
> else 예약어를 쓰지 않는다.
> else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.

현재 구조가 else 예약어를 쓰는 것과 무엇이 다른지, switch/case로 구현하는 것과 무엇이 다른가요?
`else 예약어를 쓰지 않는다`은 어떠한 학습을 위한 제약일까요? 다형성에 대해 알아보시면 좋을 듯 해요~

sinb57 commented on src/main/java/domain/Attend.java:
```diff
@@ -0,0 +1,63 @@
+
+    public static Attend of(final String day, final String time) {
+        return new Attend(DateUtil.parseDate(day), DateUtil.parsetime(time));
+    }
```
왜 도메인 객체에서 파싱해야 하나요?
이를 위해 String 타입을 인자로 받아야 할 필요가 있을까요?

문자열을 파싱하는 것은 View에 종속된 책임 아닐까요?
입력 형식이 변경되었을 때 도메인 객체가 영향 받는 것이 옳을까요?

도메인 객체가 Util에 의존하는 것은 옳을까요?
Util은 어떠한 책임을 지닌 객체인가요?

sinb57 commented on src/main/java/domain/Attend.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public boolean isDayEqual(final int day) {
+        return DateUtil.isDayEqual(day, date);
```
이 객체에서만 사용하는데, 해당 로직을 Util로 만든 이유가 있나요?

sinb57 commented on src/main/java/util/DateUtil.java:
```diff
@@ -0,0 +1,70 @@
+
+    public static boolean isDayOff(Attend attend) {
+        return isDayOff(attend.getDay());
+    }
```
Attend 객체가 책임을 지니도록 할 수 없을까요? 그렇다면 getter를 외부에서 사용할 이유가 없어요

sinb57 commented on src/main/java/util/DateUtil.java:
```diff
@@ -0,0 +1,70 @@
+    public static boolean isDayOff(int day) {
+        LocalDate targetDate = LocalDate.of(2024, 12, day);
+        return isDayOff(targetDate);
+    }
```
내부에서만 사용하는 메서드는 private으로 닫아주세요~

sinb57 commented on src/main/java/util/DateUtil.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DateUtil {
```
객체들이 지녀야 하지 않을까 싶은 메서드들이 여럿 보이네요.
책임을 옮길 객체들이 있지 않나요? 오히려 이 곳에 위치하기에 getter를 사용하게 되네요.

DateUtil의 역할은 무엇인가요? 플린트가 생각하는 Util은 어떠한 책임을 지니나요?

sinb57 commented on src/main/java/domain/AttendCount.java:
```diff
@@ -0,0 +1,25 @@
+        }
+        if (totalAbsenceCount >= 2) {
+            return WarningStatus.WARNING;
+        }
```
> else 예약어를 쓰지 않는다.
> else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.

이 부분도 AttendanceController와 마찬가지로 수정해주세요~

sinb57 commented on src/main/java/domain/AttendCount.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    public long calculateRank() {
+        return absence * 3 + late;
```
`3`은 어떠한 의미를 지녔나요?

sinb57 commented on src/main/java/domain/AttendStatus.java:
```diff
@@ -0,0 +1,18 @@
+            return LATE;
+        }
+        return ATTEND;
+    }
```
enum의 값과 조건문의 응집도를 높일 수는 없을까요? enum의 장점! 다형성을 활용해보세요~

sinb57 commented on src/main/java/domain/AttendanceResults.java:
```diff
@@ -0,0 +1,29 @@
+                .count();
+        long absenceCount = attendanceResults.stream()
+                .filter(attendanceResult -> attendanceResult.attendStatus() == AttendStatus.ABSENCE)
+                .count();
```
어떻게 하면 중복을 제거할 수 있을까요?

sinb57 commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,79 @@
+
+            // 찾은 객체를 토대로 list에서 삭제
+            removeByAttend(before);
+        }
```
어떻게 하면 주석 없이 의미를 표현할 수 있을까요?

sinb57 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,70 @@
+        String input = scanner.nextLine();
+        validateEmptyInput(input);
+        return input;
+    }
```
구조가 반복되네요! 어떻게 하면 중복을 제거할 수 있을까요???

sinb57 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,124 @@
+        }
+        if (attendStatus == AttendStatus.ABSENCE) {
+            return "결석";
+        }
```
if-else와 다르지 않네요. 어떻게 하면 if 없이 구현할 수 있을까요?

sinb57 commented on src/test/java/AttendanceBookTest.java:
```diff
@@ -0,0 +1,300 @@
+        Assertions.assertThat(attendanceBook.findByName(name).attends).hasSize(1);
+    }
+
+    @Test
+    @DisplayName("이름이 존재하지 않는 크루 출석시 예외")
+    void test524523432() throws Exception {
```
테스트명에도 신경 부탁드려요 👀

sinb57 commented on src/test/java/AttendanceBookTest.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+
+    @Test
+    void 주말_출석_저장_시도하면_예외() throws Exception {
```
DisplayName을 사용하기도 하고 한글명을 사용하기도 하는데, 기준이 무엇인가요?

sinb57 commented on src/test/java/AttendanceBookTest.java:
```diff
@@ -0,0 +1,300 @@
+public class AttendanceBookTest {
+
+    @Test
+    void 평일_출석_저장_테스트() throws Exception {
```
`throws Exception`을 붙인 이유가 궁금해요

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,106 @@
+
+    }
+
+    private AttendanceBook loadAttendanceBook(List<String> data) {
+        AttendanceBook attendanceBook = new AttendanceBook();
+        for (String row : data) {
+            final String name = parseName(row);
+            final Attend attend = parseAttend(row);
+            attendanceBook.registerName(name);
+            attendanceBook.attend(name, attend);
+        }
+        return attendanceBook;
+    }
+
+    private Attend parseAttend(String row) {
+        String dateTime = row.split(",")[1];
+        String day = dateTime.substring(8, 10);
+        String time = dateTime.substring(11);
+        return Attend.of(day, time);
+    }
+
+    private String parseName(String row) {
+        return row.split(",")[0];
+    }
+
+    private List<String> readCsv() {
+        URL fileURL = AttendanceController.class.getClassLoader().getResource(CSV_PATH);
+        return FileUtil.readFile(fileURL);
+    }
```
해당 역할을 책임지는 객체를 만들어보면 어떨까요?

sinb57 reviewed (changes requested):
안녕하세요 플린트~ 출석 미션 리뷰어를 맡은 이프입니다! 만나서 반가워요ㅎ

구현해주신 내용 확인해보니, 코드 전반에 걸쳐 같은 피드백을 반복하더라고요.
그래서 주요 피드백만 먼저 정리해 드릴게요~ 반영해주시면 이어서 추가 피드백 드릴게요!

- 내부에서 사용하는 메서드와 필드는 private으로 닫아주세요.
- if - else 의 사용과 다르지 않으므로, 다형성에 대해 알아보고 수정해주세요.
- Util에 존재하는 행위를 적절한 객체로 위임해주세요.
- 미션에 주어진 `프로그래밍 요구 사항`이 무엇을 위함인지 의식하며 지켜주세요.

sinb57 commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,117 @@
+    private static final LocalTime ABSENCE_TIME = LocalTime.of(10, 30);
+    private static final LocalTime END_TIME = LocalTime.of(23, 0);
+
+    public Map<String, Attends> map;
```
자료구조에 의존하는 변수명은 지양해주세요~

sinb57 reviewed (changes requested):
> 컨트롤러에서 도메인에 객체를 생성한 다음 객체를 메서드에 넘기는 방법과 사용할 도메인 내에서 필요한 데이터를 넘기고 메서드 내에서 필요한 객체를 생성하는 방법 중 어떤 방법이 더 좋을지 고민이 많았습니다. 개인적으로는 tdd를 진행할 때, 객체를 생성해서 메서드에 넘겨주는 방법이 테스트 코드를 작성할 떄 더 편한 느낌이었습니다. 해당 요소에 대해서 이프의 생각이 궁금합니다!

상황에 따라 선택하는 방법이 달라서 정답은 없지만, 대체로 객체간의 관계에 유의하여 구현해요~

```java
    Item item = new Item("사과", 1);
    Shop.addItem(item);
```
위 예시에 기반하여 설명드리자면, 아마도 `Shop.addItem`의 인자가 Item 타입인 이유는 Item 객체가 넘어와야만 하는 상황이겠죠? 그런데 내부 구현이 단순히 name(string)과 count(int)를 받아와 처리하는 것이라면 Item 객체로 넘기지 않을 수 있어요. 이 경우는 인자로 넘어가는 name과 count가 Item 객체와 연관이 없어야 해요. `(name, count) != Item`이어야 하는거죠. 만약 서로다른 Item 객체에 대해 한 객체의 name과 또 다른 객체의 count를 넘겨야 하는 상황이라면?? `Shot.addItem`의 인자로 2개의 서로다른 객체를 전달하는 방식을 택할거예요.

설명이 다소 난해한데, 쉽게 한 문장으로 설명드릴게요.
객체의 메시징에 집중하세요. 메서드의 시그니처를 어떻게 표현할지에 집중하세요.

jhan0121 committed:
refactor: CSV 파일 읽기 기능 객체 분리

jhan0121 committed:
refactor: 출석 상태 메시지 분리

jhan0121 committed:
refactor: 경고 대상자 판정 및 대상자 메시지 분리

jhan0121 committed:
feat: OutputView 주석 수정

jhan0121 committed:
refactor: OutputView 출력 형식 수정

jhan0121 committed:
refactor: 컨트롤러 기능 별 메서드 분리

jhan0121 committed:
refactor: 경고 판정 객체 내 대상 판정 메서드 구현

jhan0121 committed:
refactor: DateUtil 내 주말 판정 기능 및 평일 일수 계산 로직 분리

jhan0121 committed:
refactor: Attend 내 메서드 구조 수정

- 멤버변수 private 설정
- time 기반 생성 메서드명 변경
- dateUtil 내 dayOff, isTimeOff 메서드 분리
- isBefore, isEqual, isAfter 로직 수정
- getHour, getMinute 에서 time이 null일 때 문제 수정
- AttendTest 수정 및 추가

jhan0121 committed:
refactor: AttendanceBook 구조 및 메서드 수정

- 멤버 변수 private 설정 및 이름 수정
- checkWarningCrew 인덴트 구조 수정
- 메서드 위치 수정
- 테스트 코드 구조 수정 및 추가

jhan0121 committed:
refactor: 출석 결과 판정 기능 중복 로직 수정

jhan0121 committed:
test: AttendResults 테스트 코드 수정

jhan0121 committed:
docs: csv 파일 데이터 수정

jhan0121 committed:
refactor: 출석 수 계산 판정 로직 구조 수정

- 테스트 코드 구조 추가 및 수정

jhan0121 committed:
refactor: 출석 상태 판정 로직 구조 수정

- 테스트 구조 수정 및 추가

jhan0121 committed:
refactor: InputView 입력 메서드 리턴 값 수정

jhan0121 committed:
refactor: Attends 구조 수정

- 멤버 변수 private 설정
- 메서드 순서 수정
- 테스트 코드 구조 수정 및 추가

jhan0121 committed:
refactor: DateUtil 제거

jhan0121 committed:
refactor: getYesterday 메서드 삭제

jhan0121 committed:
refactor: String 기반 생성 메서드 삭제

jhan0121 committed:
refactor: 출석 결과 출력 구조 수정

jhan0121 committed:
refactor: 컨트롤러 인덴트 및 메서드 길이 요구사항 반영

jhan0121 commented:
안녕하세요 이프! 소중한 피드백 감사합니다!
처음으로 tdd를 적용해서 미션을 수행하다 보니 객체 구조 나 테스트 코드 구조 와 같은 기본적인 요구 사항들을 지키지 못해 아쉬웠습니다.

이번 리팩터링에서 반영한 요소는 크게 5가지입니다.
- 인덴트 및 메서드 길이 요구사항 반영하기
- 멤버 변수 private 변경 및 변수명 수정
- Util 객체에 있는 메서드 위치 수정 및 Util 패키지 제거
- if 문 구조 수정하기(Enum의 다형성 적용)
- 테스트 코드 구조 수정

이번 피드백을 통해 `Enum을 이용한 다형성`에 대해서 처음으로 알게 되었고 리팩터링에 적용해보았습니다! 

궁금한 점으로 1가지 있습니다.

view에 있는 if문 구조를 수정하기 위해 enum을 적용해보았습니다.(view 패키지 내 AttendMessage, WarningMessage) 해당 객체들은 view에 String을 제공하는 역할만 하기 때문에 view 패키지에 넣어도 되지 않을까? 라는 생각이 들어 view 패키지에 넣었습니다. 그런데 기능에 대해서 역할을 가지고 있다면 domain으로 봐야 할 것 같은데 해당 객체는 도메인이 맞는지 의문이 들었습니다. `조건에 맞는 문자열을 찾기 위한 로직`을 가진 객체도 domain으로 봐야 할까요? 이프의 생각이 궁금합니다!

sinb57 commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,75 @@
+        for (int day : existAttendDay) {
+            result.add(findByDay(day));
+        }
+        return result;
```
stream 활용할 수 있어보이네요!

jhan0121 replied:
stream으로 구현해서 리턴하도록 재구성하였습니다!
```java
public List<Attend> getAttends(List<Integer> dayOfWeek) {
        List<Attend> result = new ArrayList<>();
        dayOfWeek.stream()
                .filter(this::hasDayEqualsAttend)
                .forEach(day -> result.add(findByDay(day)));
        return result;
    }
```

sinb57 commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,75 @@
+                .filter(target::isDayEqual)
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("출석을 찾는데 실패하였음"));
+    }
```
validateDuplicate 메서드를 보면 생성자에서 사용하고 있네요. 생성자와 해당 메서드의 사이의 거리가 멀다보니, 생성자에서 시작하여 호출 메서드를 들여다보기 위해서는 비교적 먼 거리를 이동하여 메서드를 살펴보게 됩니다. 크진 않지만 시간과 노력이 소모되다 보니, 수직거리를 고려해주시면 좋아요! 시간 괜찮을 때 "클린코드(p.101)"를 읽어보세요~

jhan0121 replied:
해당 메서드 뿐만 아니라 전체적으로 연관이 있는 메서드끼리 가까운 거리에 존재하도록 재구성하였습니다!

```java
    public void addAttend(Attend attend) {
        validateDuplicate(attend);
        attends.add(attend);
    }
// 연관성 높은 메서드 간 거리 조정
    private void validateDuplicate(Attend targetAttend) {
        if (hasDayEqualsAttend(targetAttend)) {
            throw new IllegalArgumentException("같은 날짜에 출석할 수 없다.");
        }
    }
```

sinb57 replied:
다른 객체들도 반영 해주세요~ㅎ

jhan0121 replied:
아래의 객체들도 피드백 반영하였습니다!

- Attend
- AttendanceBook
- AttendanceController
- Attends
- InputView
- OutputView

sinb57 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,98 @@
+
+    private void runCommand(String command, AttendanceBook attendanceBook) {
+        try {
+            registerAttend(command, attendanceBook);
+            editAttend(command, attendanceBook);
+            searchAttendance(command, attendanceBook);
+            searchWarningCrews(command, attendanceBook);
```
이 방식은 2개의 단점을 지니고 있어요.
- 코드를 읽고 파악하는 입장에서는, 나열된 4개의 메서드가 모두 실행되는 것으로 보여요.  출석을 등록하고 수정하고 조회하고 위험자를 확인한다고 말이죠. 하지만 우리의 의도는 그렇지 않아요. 내부 구현을 들여다보면 command에 따라 하나의 행위가 선택되어 실행되는 것을 볼 수 있어요. 즉 4개의 행위가 같이 실행되는 것이 아니라 하나의 행위가 선택되길 바라는 거죠. 그런데 이러한 의미 파악은 내부 구현을 들여다 봐야지만 알 수 있어요. 과연 옳을까요? 이는 코드 파악에 큰 장애물이 돼요. 외부에서는 메서드의 시그니처만을 바라보고 사용목적과 의도를 추측하여 사용합니다. 신뢰를 바탕으로 진행되죠. 하지만 위와 같이 밖으로 드러나는 것과 내부 구현이 다르다면, 이 프로젝트의 모든 메서드는 앞으로 내부 구현을 들여다보고 이해해야만 가져다 사용할 수 있게 됩니다.
- 그리고 4개의 메서드를 풀어 써보면 if문의 나열과 다르지 않아요. 로직의 흐름이 if문의 나열과 다르지 않다보니, if-else보다 성능이 좋지 못해요.

어떻게 하면 개선할 수 있을까요?

jhan0121 replied:
Map을 활용해서 문제를 해결해 보았습니다! 커맨드로 사용되는 문자열은 `Command` 라는 Enum을 활용해서 값에 의미를 부여했습니다. 각 `Command`에 맞는 실행 메서드를 `Consumer`로 구성하여 value로 삽입해주었습니다.

```java
private final Map<Command, Consumer<AttendanceBook>> commands;

    public AttendanceController(InputView inputView, OutputView outputView) {
        this.inputView = inputView;
        this.outputView = outputView;
        this.commands = Map.of(
                Command.ATTEND, this::registerAttend,
                Command.EDIT, this::editAttend,
                Command.SEARCH_ATTEND, this::searchAttendance,
                Command.SEARCH_WARNING_CREW, this::searchWarningCrews
        );
    }
```

sinb57 commented on src/main/java/domain/Attend.java:
```diff
@@ -0,0 +1,103 @@
+
+    public boolean isDayOff() {
+        return this.date.getDayOfWeek().getValue() >= DayOfWeek.SATURDAY.getValue()
+                || this.date.getDayOfMonth() == Current.CHRISTMAS;
```
> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면 위와 같이 나와 있네요. 본 시스템은 12월을 한정으로 하였으나, 시범 운영인 만큼 모든 월을 대상으로 확장될 가능성이 높습니다.

현재의 코드 상으로는 모든 월의 25일을 크리스마스라고 볼 텐데, 이 경우 변경이 작지 않을 것 같네요. 확장 가능성을 고려하여 개선해보시겠어요?

jhan0121 replied:
`Holiday`라는 enum을 구성해서 `토요일`, `일요일` 외에 `공휴일`을 관리할 수 있는 객체를 구성했습니다.  공휴일 판정 메서드인 `isHoliday`도 추가하여 Attend에서 활용하도록 로직을 수정했습니다.

## Holiday
```java
public enum Holiday {
    CHRISTMAS("12-25");

    private final String date;

    Holiday(String date) {
        this.date = date;
    }

    public static boolean isHoliday(LocalDate targetDate) {
        return Arrays.stream(Holiday.values())
                .map(holiday -> Holiday.parseDate(holiday.date))
                .map(LocalDate::parse)
                .anyMatch(targetDate::equals);
    }
}
```

## Attend
```java
    public boolean isDayOff() {
        return this.date.getDayOfWeek().getValue() >= DayOfWeek.SATURDAY.getValue()
                || Holiday.isHoliday(this.date);
    }
```

sinb57 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private String formatAttendAndStatus(Attend attend, AttendStatus attendStatus) {
+        String date = attend.formatDate(DateTimeFormatter.ofPattern("MM월 dd일 E요일"));
```
메서드를 호출할 때마다 DateTimeFormatter 객체가 생성될 텐데요.
"이펙티브 자바"의 "아이템 6. 불필요한 객체 생성을 피하라" 내용을 읽어 보길 바래요~

jhan0121 replied:
`DateFormat`이라는 enum 객체를 구성하였습니다.
 `DateTimeFormatter`를 제공하는 기능을 추가해서 메서드를 호출할 때마다 
반복되는 객체가 생성되지 않도록 구성하였습니다.

```java
public enum DateTimeFormat {
    DATE(DateTimeFormatter.ofPattern("MM월 dd일 E요일")),
    TIME(DateTimeFormatter.ofPattern("HH:mm"));

    private final DateTimeFormatter dateTimeFormatter;

    DateTimeFormat(DateTimeFormatter dateTimeFormatter) {
        this.dateTimeFormatter = dateTimeFormatter;
    }

    public DateTimeFormatter getDateTimeFormatter() {
        return dateTimeFormatter;
    }
}
```

sinb57 commented on src/main/java/domain/WarningStatus.java:
```diff
@@ -0,0 +1,30 @@
+package domain;
+
+public enum WarningStatus {
+    CLEAR {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount < 2;
+        }
+    },
+    WARNING {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount == 2;
+        }
+    },
+    INTERVIEW {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount >= 3 && totalAbsenceCount <= 5;
+        }
+    },
+    EXPEL {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount > 5;
+        }
+    };
```
매직넘버가 보이네요!
- 기획서를 보기 전에는 2가 무엇이고 3은 무엇인지 추측하기 어려웠어요.
- 그리고 기준이 변경되면, 최소 2곳을 수정해야 하는데 실수로 누락할 수도 있어보여요.

jhan0121 replied:
각 매직 넘버를 상수화해서 의미 부여를 해준 다음에 판정식을 수정하였습니다.

```java
CLEAR((totalAbsenceCount) -> totalAbsenceCount < WarningStatus.LATE_BOUNDARY),
    WARNING((totalAbsenceCount) -> totalAbsenceCount == WarningStatus.LATE_BOUNDARY),
    INTERVIEW((totalAbsenceCount) -> totalAbsenceCount > WarningStatus.LATE_BOUNDARY
            && totalAbsenceCount <= WarningStatus.ABSENCE_BOUNDARY),
    EXPEL((totalAbsenceCount) -> totalAbsenceCount > WarningStatus.ABSENCE_BOUNDARY);

    private static final int LATE_BOUNDARY = 2;
    private static final int ABSENCE_BOUNDARY = 5;

```

sinb57 commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,122 @@
+        this.attendsPerCrew = new HashMap<>();
+    }
+
+    public void registerName(String name) {
+        if (attendsPerCrew.containsKey(name)) {
+            return;
+        }
+        attendsPerCrew.put(name, new Attends(new ArrayList<>()));
+    }
```
putIfAbsent 와 computeIfAbsent 에 대해 찾아보면 좋을 듯 해요!

jhan0121 replied:
`putIfAbsent`를 적용하여 코드 구조를 단순화하였습니다.

```java
public void registerName(String name) {
        attendsPerCrew.putIfAbsent(name, new Attends(new ArrayList<>()));
    }
```

sinb57 commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,122 @@
+        return result;
+    }
+
+    private void validateAttendableTime(Attend attend) {
+        if (attend.isTimeOff(START_TIME, END_TIME)) {
+            throw new IllegalArgumentException("운영 시간 외에는 출석할 수 없음");
+        }
+    }
+
+    private void validateAttendableDay(Attend attend) {
+        if (attend.isDayOff()) {
+            throw new IllegalArgumentException("쉬는날은 출석할 수 없음");
+        }
+    }
```
운영 시간이라는 값 객체를 만들면 책임을 위임할 수 있어보여요!

jhan0121 replied:
`OperationTime` 객체를 구성하여 AttendanceBook에 있던 상수들과 판정 메서드를 이동하였습니다.

```java
START_TIME(LocalTime.of(8, 0)),
    LATE_TIME(LocalTime.of(10, 5)),
    ABSENCE_TIME(LocalTime.of(10, 30)),
    END_TIME(LocalTime.of(23, 0));

...

    public static void validateAttendableDay(Attend attend) {
        if (attend.isDayOff()) {
            throw new IllegalArgumentException("쉬는날은 출석할 수 없음");
        }
    }

    public static void validateAttendableTime(Attend attend) {
        if (attend.isTimeOff(START_TIME.time, END_TIME.time)) {
            throw new IllegalArgumentException("운영 시간 외에는 출석할 수 없음");
        }
    }
```

현재 구현에서 `Attend` 객체에  attend와 주어진 시간(여기에서는 운영 시작시간, 운영 종료 시간)을 메시지로 보내서 그 결과를 boolean으로 받도록 구성이 되어있습니다. attend가 아닌 `LocalTime`을 직접 이용해보는 방법도 생각해 보았지만 그렇게 된다면 `Attend`에서 getter를 구현해야 해서 고민이 많았습니다. 

제가 생각했던 attend의 역할로 `출석 날짜와 시간을 가지고 외부에서 주어진 정보들(날짜, 시간)을 토대로 어떻게 구성되어 있는지(같은지, 이전인지, 이후인지) 알려주는 역할`이라고 생각하여 구성하였습니다. 따라서 getter를 사용하지 않고 `Attend` 객체에 필요한 정보를 제공하여 그 결과를 반환해주는 방식으로 구현을 진행했었습니다. 

따라서 `OperationTime`에 존재하는 valid 관련 메서드에서 attend를 매개변수로 받도록 설정했습니다.

sinb57 commented on src/main/java/domain/AttendanceResult.java:
```diff
@@ -0,0 +1,4 @@
+package domain;
+
+public record AttendanceResult(Attend attend, AttendStatus attendStatus) {
+}
```
로직이 없는 클래스네요! 목적이 뭔가요?

jhan0121 replied:
`AttendanceResult`의 역할로 OutputView에서 출력을 위해 출석 현황(날짜, 시간, 출석 상태)를 전달하기 위한 목적으로 구성했기 때문에 로직 없이 구성했습니다.

sinb57 replied:
출력을 위한 DTO 군요! DTO를 도메인 객체에서 생성하고 있는데, 도메인이 DTO에 의존해도 괜찮을까요?

jhan0121 replied:
도메인이 dto에 의존하게 된다면 view의 요구사항이 변경될 때마다 도메인의 로직도 변경이 되어야 하기 때문에 적절하지 않다고 생각하였습니다.

해당 문제를 해결하기 위해서 `AttendanceResult`는 view에 전달하지 않는 `attend와 attendStatus를 저장하는 역할`로 재구성하였습니다. `AttendanceResult`에는 제시한 출석 상태와 같은지 확인하는 로직을 추가한 다음, `AttendanceResults`에서 사용하도록 하였습니다.

### AttendanceResult
```java
public record AttendanceResult(Attend attend, AttendStatus attendStatus) {
    public boolean isEqualStatus(AttendStatus attendStatus) {
        return this.attendStatus == attendStatus;
    }
}
```
### AttendanceResults
```java
private long countStatus(AttendStatus attendStatus) {
        return attendanceResults.stream()
                .filter(attendanceResult -> attendanceResult.isEqualStatus(attendStatus))
                .count();
    }
```

 view에서는 `AttendResultDto`를 이용하도록 하여 view에서 필요한 데이터를 넘겨주는 dto를 설정하였습니다.

### AttendResultDto
```java
public record AttendResultDto(String name, List<AttendanceResult> attendanceResults, AttendCount attendCount) {
}
```

### OutputView
```java
public void printAttendanceResult(AttendResultDto attendResultDto) {
    // DTO에 저장된 정보 추출
    String name = attendResultDto.name();
    List<AttendanceResult> attendResult = attendResultDto.attendanceResults();
    AttendCount attendCount = attendResultDto.attendCount();
    // 각 변수를 기반으로 출력
    System.out.printf("이번 달 %s의 출석 기록입니다.%n%n", name);
    printAttendanceStatus(attendResult);
    printAttendCount(attendCount);
    printWarningMessage(attendCount);
}
```

sinb57 commented on src/main/java/domain/AttendReader.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static final String CSV_PATH = "attendances.csv";
+
+    public AttendanceBook loadAttendanceBook() {
```
단위 테스트가 불필요한가요? 파일이 잘못된 포맷으로 작성됐어도 동작하면 어떡하죠?

jhan0121 replied:
해당 객체에서 Csv 파일을 읽는 역할은 CsvReader로 분리하였고, 파일 포맷이 정상인 경우와 잘못된 케이스들을 기반으로 단위 테스트 클래스(CsvReaderTest, AttendReaderTest)를 구성하였습니다.

## CsvReaderTest
- 정상 케이스
- csv 분리자로 정상적으로 구분되지 않은 경우
- datetime 열에서 datetime이 잘못된 형식인 경우
- datetime 열에서 date가 잘못된 형식인 경우
- datetime 열에서 time이 잘못된 형식인 경우

## AttendReaderTest
- 정상 케이스
- csv 분리자로 정상적으로 구분되지 않은 경우

sinb57 commented on src/main/java/domain/AttendReader.java:
```diff
@@ -0,0 +1,51 @@
+
+    private List<String> readFile(URL fileURL) {
+        try (BufferedReader reader = Files.newBufferedReader(Paths.get(fileURL.toURI()))) {
+            return reader.lines().skip(1).toList();
```
왜 첫줄을 무시하는지 의미부여하면 좋겠습니다.

jhan0121 replied:
해당 기능을 메서드로 별도 분리하여 어떤 역할을 하는지 표현하였습니다.

```java
    private static List<String> readFile(URL fileURL) {
        try (BufferedReader reader = Files.newBufferedReader(Paths.get(fileURL.toURI()))) {
            return removeTitleRow(reader);
        } catch (IOException | URISyntaxException e) {
            throw new IllegalStateException(String.format("경로 문제: %s", fileURL.getPath()));
        }
    }

    private static List<String> removeTitleRow(BufferedReader reader) {
        return reader.lines().skip(1).toList();
    }
```

sinb57 commented on src/main/java/domain/AttendReader.java:
```diff
@@ -0,0 +1,51 @@
+    private Attend parseAttend(String row) {
+        String dateTime = row.split(",")[1];
+        String[] dateTimeParse = dateTime.split(" ");
+        return Attend.of(LocalDate.parse(dateTimeParse[0].strip()), LocalTime.parse(dateTimeParse[1].strip()));
```
dateTimeParse의 0,1이 무엇을 의미하는지 의미를 나타내주세요

jhan0121 replied:
매직 넘버들을 상수로 변환하여 적용하였습니다. 위에서 말씀하신 사항 외에도 `csv 분리자`, `date-time 분리자`, `전체 행 개수`, `date-time 분리 개수`, `이름에 해당하는 인덱스`, `날짜에 해당하는 인덱스`, `시간에 해당하는 인덱스` 또한 상수로 적용하였습니다.

```java
    private static final String CSV_DELIMITER = ",";
    private static final String DATE_TIME_DELIMITER = " ";
    private static final int TOTAL_ROW_COUNT = 2;
    private static final int TOTAL_DATE_TIME_COUNT = 2;
    private static final int NAME_INDEX = 0;
    private static final int DATE_TIME_INDEX = 1;
    private static final int DATE_INDEX = 0;
    private static final int TIME_INDEX = 1;
...
public static Attend parseAttend(String row) {
...
        String dateTime = split[DATE_TIME_INDEX];
        String[] dateTimeParse = splitTime(dateTime);
...
        return Attend.of(parseDate(dateTimeParse[DATE_INDEX].strip()), parseTime(dateTimeParse[TIME_INDEX].strip()));
    }
```

여기서 궁금한 점이 있습니다.

매직 넘버들을 상수로 변환하다보니 상수의 개수가 많은 것 같아 따로 보관하는 객체를 만들어야 할 지 고민이 있었습니다. 하지만 해당 값들은 `CsvReader`에서 밖에 사용하지 않아서 별도의 객체로 분리하게 된다면 오히려 코드를 이해하는데 움직이는 거리가 멀어진다고 생각이 들었습니다. 

상수의 개수에 따른 객체 분리에 대해서 이프 만의 기준이 있는지 알고 싶습니다!

sinb57 replied:
흔히 매직넘버를 없애기 위해 상수로 처리하곤 해요. 그치만 이 경우는 직접 겪으신 듯 상수처리가 오히려 단점으로 다가옵니다. 제가 의미 부여를 요구하면서 바랬던 것은 `dateTimeParse[0].strip()`에 대한 의미 부여 였어요. 이를 변수처리하는 것만으로도 충분한 의미가 부여된다고 생각해요.

sinb57 commented on src/main/java/domain/Attend.java:
```diff
@@ -0,0 +1,103 @@
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+
+public class Attend {
+
+    private final LocalDate date;
+    private final LocalTime time;
```
`출석 일시`로 보이네요! 생성 시 검증할 만한 조건은 없을까요?

jhan0121 replied:
현재 구성에서 Attend는 출석 정책과 상관 없는 정보 만을 기반으로 검증을 진행해야 하지 않을까? 라는 생각이 들었습니다.

Attend의 경우에서 생성 시 검증할 만한 조건으로는 `date가 null이 아닌지` 정도로 생각하였습니다. `time`의 경우에는 `CSV파일에 존재하지 않는 날짜의 출석 데이터`는 `결석`으로 간주하도록 구성되어있기 때문에 `null`이 될 수 있다고 판단했습니다.

그리고 현재 구성에서 Attend에 있는 정보 만으로는 `time이 운영 시간 사이인지`, `date가 쉬는 날 인지`에 대한 검증은 출석 정책을 가지고 있는 OperationTime에 존재하는 데이터를 기반으로 검증을 진행해야 하지 않을까? 라는 생각이 들었습니다. 왜냐하면 사용자가 원하는 정책에 따라서 적용되는 `출석 가능 일자`가 달라질 수 있기 때문입니다.(예: 어떤 곳에서는 월 ~ 토까지 출석할 수 있다면? OR 운영 시간이 12:00 ~ 23:59라면?)

따라서 `time이 운영 시간 사이인지`, `date가 쉬는 날 인지`에 대해서는 `OperationTime`을 기반으로 하여 AttendanceBook에서 확인하는 것이 적절하다고 판단하였습니다.

```java
private Attend(LocalDate date, LocalTime time) {
        validateDateIsNotNull(date);
        this.date = date;
        this.time = time;
    }

    private void validateDateIsNotNull(LocalDate date) {
        if (date == null) {
            throw new IllegalArgumentException("date는 null이 될 수 없음");
        }
    }
```

sinb57 commented on src/test/java/AttendanceBookTest.java:
```diff
@@ -0,0 +1,353 @@
+        AttendanceBook attendanceBook = new AttendanceBook();
+        attendanceBook.registerName(name);
+        List<Attend> expectAttend = List.of(
+                Attend.of(LocalDate.of(2024, 12, 2), LocalTime.of(10, 0)),
+                Attend.of(LocalDate.of(2024, 12, 3), LocalTime.of(10, 0)),
+                Attend.of(LocalDate.of(2024, 12, 4), LocalTime.of(10, 0)),
+                Attend.of(LocalDate.of(2024, 12, 5), LocalTime.of(10, 6)),
+                Attend.of(LocalDate.of(2024, 12, 6), LocalTime.of(10, 6)),
```
테스트를 위한 팩토리 메서드를 생성하여 테스트의 작성 부담을 줄이는 것도 좋은 방법입니다.

jhan0121 replied:
테스트에서 사용할 Attend 생성용 메서드를 작성해서 중복되는 코드를 줄였습니다.

```java
    private List<Attend> createAttendsUntilDay(int day) {
        List<Attend> result = new ArrayList<>();
        IntStream.range(1, day).forEach(i -> result.add(Attend.fromDay(i)));
        return result;
    }

    private List<Attend> createAttendsByTargetDay(List<Integer> days) {
        List<Attend> result = new ArrayList<>();
        days.forEach(day -> result.add(Attend.fromDay(day)));
        return result;
    }

    @Test
    @DisplayName("출석 대상 날짜들이 들어있는 리스트를 토대로 출석 객체들을 반환한다")
    void shouldReturnAttendsByDayOfWeek() {
...
        List<Attend> attendsInitValue = createAttendsUntilDay(Current.TODAY.getDay());
...
        List<Attend> actual = createAttendsByTargetDay(dayOfWeek);
...
    }
```

그런데 이프가 말씀하신 `테스트를 위한 팩토리 메서드를 생성`한다는 의미를 제대로 적용한 것인지는 잘 모르겠습니다. 
만약 적용한 방법이 아닌 다른 의도로 말씀하신 것이라면 어떤 방식을 적용해야 했는지 알 수 있을까요?

sinb57 replied:
간단하게 생각해서, `createAttend(year, month, day, hour, minute)`를 생각했어요. 그치만 테스트 픽스처를 만드는 행위로, 플린트가 제안 준 방법도 좋아요

sinb57 commented on src/main/java/view/AttendMessage.java:
```diff
@@ -0,0 +1,37 @@
+
+public enum AttendMessage {
+
+    ATTEND("출석") {
+        @Override
+        public boolean match(AttendStatus attendStatus) {
+            return attendStatus == AttendStatus.ATTEND;
+        }
+    },
+    LATE("지각") {
+        @Override
+        public boolean match(AttendStatus attendStatus) {
+            return attendStatus == AttendStatus.LATE;
+        }
+    },
```
Enum을 활용하셨군요ㅎㅎ 잘해주셨습니다~

jhan0121 replied:
이번 기회에 Enum을 활용한 다형성에 대해서 알 수 있었습니다!

sinb57 commented on src/main/java/view/AttendMessage.java:
```diff
@@ -0,0 +1,37 @@
+            return attendStatus == AttendStatus.LATE;
+        }
+    },
+    ABSENCE("결석") {
+        @Override
+        public boolean match(AttendStatus attendStatus) {
+            return attendStatus == AttendStatus.ABSENCE;
+        }
+    };
```
match 메서드의 로직이 모두 단순일치여부 확인이네요. 그렇다면 `ABSENCE("결석", AttendStatus.ABSENCE)`라고 해놓고서 활용하면 어떤가요?
`OutputView#formatAttendStatus` 로직을  옮겨올 수 있겠네요

jhan0121 replied:
enum 구조를 말씀하신 피드백을 기반으로 재구성하였습니다.

OutputView에 존재하던 `formatAttendStatus`메서드 또한 이동시켜 역할을  재분배하였습니다. 

```java
    ATTEND("출석", AttendStatus.ATTEND),
    LATE("지각", AttendStatus.LATE),
    ABSENCE("결석", AttendStatus.ABSENCE);

    private final String message;
    private final AttendStatus attendStatus;

...

    public static String formatAttendStatus(AttendStatus attendStatus) {
        return Arrays.stream(AttendMessage.values())
                .filter(attendMessage -> attendMessage.attendStatus == attendStatus)
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("출석 상태 판정 실패"))
                .getMessage();
    }
...
```

sinb57 commented on src/main/java/domain/WarningStatus.java:
```diff
@@ -0,0 +1,30 @@
+package domain;
+
+public enum WarningStatus {
+    CLEAR {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount < 2;
+        }
+    },
+    WARNING {
+        @Override
+        public boolean match(long totalAbsenceCount) {
+            return totalAbsenceCount == 2;
+        }
+    },
```
메서드를 오버라이딩하기 위해 같은 구조를 반복 작성/수정해야 하는 불편함이 있네요. 함수형 인터페이스를 활용해보면 어떨까요?

jhan0121 replied:
AttendStatus와 WarningStatus에 함수형 인터페이스를 적용하여 중복되는 코드를  제거하였습니다.  또한 판정 로직도 enum 객체로 이동시켜 역할을 재분배하였습니다. 

### AttendStatus
```java
ATTEND((attend) -> attend.hasTime() && attend.isBefore(OperationTime.LATE_TIME.getTime())),
    LATE((attend) -> attend.hasTime()
            && (attend.isEqual(OperationTime.LATE_TIME.getTime()) || attend.isAfter(OperationTime.LATE_TIME.getTime()))
            && attend.isBefore(OperationTime.ABSENCE_TIME.getTime())),
    ABSENCE((attend) -> !attend.hasTime()
            || attend.isEqual(OperationTime.ABSENCE_TIME.getTime())
            || attend.isAfter(OperationTime.ABSENCE_TIME.getTime()));

    private final Predicate<Attend> matchCondition;
...
    public static AttendStatus findAttendStatus(Attend attend) {
        return Arrays.stream(AttendStatus.values())
                .filter(attendStatus -> attendStatus.matchCondition.test(attend))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("출석 상태 판정 실패"));
    }
```

### WarningStatus
```java
CLEAR((totalAbsenceCount) -> totalAbsenceCount < WarningStatus.LATE_BOUNDARY),
    WARNING((totalAbsenceCount) -> totalAbsenceCount == WarningStatus.LATE_BOUNDARY),
    INTERVIEW((totalAbsenceCount) -> totalAbsenceCount > WarningStatus.LATE_BOUNDARY
            && totalAbsenceCount <= WarningStatus.ABSENCE_BOUNDARY),
    EXPEL((totalAbsenceCount) -> totalAbsenceCount > WarningStatus.ABSENCE_BOUNDARY);

    private static final int LATE_BOUNDARY = 2;
    private static final int ABSENCE_BOUNDARY = 5;

    private final Predicate<Long> matchCondition;
...
    public static WarningStatus judgeWarningStatus(long totalAbsenceCount) {
        return Arrays.stream(WarningStatus.values())
                .filter(warningStatus -> warningStatus.matchCondition.test(totalAbsenceCount))
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("경고 상태 판정 실패"));
    }
```

sinb57 reviewed (changes requested):
안녕하세요 플린트~ 반영 잘 해주셨네요ㅎ 고생 많으셨습니다!!
내용 확인했고 추가 코멘트 남겼어요~ 확인 부탁드려요!

---

> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

---

> Q. view에 있는 if문 구조를 수정하기 위해 enum을 적용해보았습니다.(view 패키지 내 AttendMessage, WarningMessage) 해당 객체들은 view에 String을 제공하는 역할만 하기 때문에 view 패키지에 넣어도 되지 않을까? 라는 생각이 들어 view 패키지에 넣었습니다. 그런데 기능에 대해서 역할을 가지고 있다면 domain으로 봐야 할 것 같은데 해당 객체는 도메인이 맞는지 의문이 들었습니다. 조건에 맞는 문자열을 찾기 위한 로직을 가진 객체도 domain으로 봐야 할까요? 이프의 생각이 궁금합니다!

A. 목적이 다르다고 생각해요. 입출력을 위한 로직이라면 입출력에 존재해도 좋지 않을까요? 플린트는 domain 패키지와 view 패키지를 어떠한 기준으로 나눴나요? 나눠야 하는 이유는 무엇인가요? 비즈니스 로직과 UI 로직은 왜 분리하는 것이 좋은가요?

jhan0121 committed:
refactor: Attends 내 getAttends 메서드에 stream 적용

jhan0121 committed:
style: Attends 메소드 순서 연관 기능별로 재정렬

jhan0121 committed:
refactor: 공휴일 데이터 저장 객체 추가

jhan0121 committed:
refactor: Date 또는 Time 출력을 위한 formatter 저장 객체 추가

jhan0121 committed:
refactor: AttendMessage 판정 로직 수정

jhan0121 committed:
refactor: WarningMessage 판정 로직 수정

jhan0121 committed:
refactor: WarningStatus 판정 메서드 구조 수정:

jhan0121 committed:
refactor: Controller 커맨드 판정 기능 수정

jhan0121 committed:
refactor: Attend fromTime 로직 수정

jhan0121 committed:
refactor: 연관성 높은 메소드 위치 조정

jhan0121 commented:
안녕하세요 이프!
말씀해주신 피드백 반영해서 수정했습니다!

주요 변경점으로는 4가지가 있습니다.
- 컨트롤러 내 기능 판정 로직 Map을 활용하도록 변경
- Enum 기반 다형성 로직 수정
- formatter 를 제공하는 enum 객체 추가
- 연관된 메소드 간 위치 조정

jhan0121 committed:
refactor: 생성자 내 검증 로직 추가

jhan0121 commented:
안녕하세요 이프!
말씀해주신 피드백 반영해서 수정했습니다!

주요 변경점으로는 5가지가 있습니다.
- 컨트롤러 내 기능 판정 로직 Map을 활용하도록 변경
- Enum 기반 다형성 로직 수정
- formatter 를 제공하는 enum 객체 추가
- 연관된 메소드 간 위치 조정
- 공휴일 데이터 enum 분리

> 목적이 다르다고 생각해요. 입출력을 위한 로직이라면 입출력에 존재해도 좋지 않을까요? 플린트는 domain 패키지와 view 패키지를 어떠한 기준으로 나눴나요? 나눠야 하는 이유는 무엇인가요? 비즈니스 로직과 UI 로직은 왜 분리하는 것이 좋은가요?

A. view의 역할은 domain에 존재하는 데이터를 기반으로 출력하는 역할로, domain은 view에서 온 요청을 기반으로 데이터를 조작하는 역할로 생각하였습니다. 비즈니스 로직과 UI 로직을 분리하면 UI 또는 비즈니스 로직에 변화가 생겨도 서로 영향이 적어지게 되어 유지보수성이 좋아지기 때문입니다. 다시 생각해보니 AttendMessage, WarningMessage는 View에 위치하는 것이 더 적절하다는 생각이 들었습니다!

sinb57 commented on src/main/java/domain/Command.java:
```diff
@@ -0,0 +1,25 @@
+    public static Command judgeCommand(String command) {
+        return Arrays.stream(Command.values())
+                .filter(com -> com.command != null)
+                .filter(com -> com.command.equals(command))
```
약어 사용은 지양해주세요

jhan0121 replied:
매개변수명과 중복되다보니 약어를 사용했던 것 같습니다!

아래와 같이 매개변수명은 `commandInput`으로 변경하고 `com`을 `command`로 변경하여 약어를 제거하였습니다.

```java
    public static Command judgeCommand(String commandInput) {
        return Arrays.stream(Command.values())
                .filter(command -> command.command != null)
                .filter(command -> command.command.equals(commandInput))
                .findAny()
                .orElse(UNKNOWN);
    }
```

jhan0121 committed:
refactor: 매직 넘버 상수화

jhan0121 committed:
refactor: AttendanceBook 크루 등록 기능 로직 수정

jhan0121 committed:
refactor: AttendReader 제목 행 제거 메서드 추가

jhan0121 committed:
refactor: AttendReader 매직 넘버 상수화

jhan0121 committed:
refactor: 데이터 내 제목 행 제거 메서드 추가

jhan0121 committed:
refactor: OperationTime 객체 역할 분리

- 운영 시간 정보 및 판정 로직 분리

jhan0121 committed:
refactor: Attend 내 불필요한 로직 제거

jhan0121 committed:
refactor: Current 구조 변경

jhan0121 committed:
refactor: Holiday 구조 변경

jhan0121 committed:
feat: CsvReader 추가

jhan0121 committed:
refacor: 파일 읽기 기능 AttendReader 객체로 분리

jhan0121 committed:
refactor: AttendStatus 함수형 인터페이스 구조로 변경

jhan0121 committed:
refactor: Attend fromDay Current 상수를 사용하도록 변경

jhan0121 committed:
test: AttendReaderTest 추가

sinb57 commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public List<Attend> getAttends(List<Integer> dayOfWeek) {
+        List<Attend> result = new ArrayList<>();
+        dayOfWeek.stream()
+                .filter(this::hasDayEqualsAttend)
+                .forEach(day -> result.add(findByDay(day)));
+        return result;
```
stream의 `map()`을 사용하면 `List<Attend> result = new ArrayList<>();` 없이 곧바로 toList()가 가능해요~

jhan0121 replied:
`map()`을 적용하여 코드 구조를 단순화하였습니다! 또한 중복되는 메서드 호출이 존재하는 것을 확인하여 중복 되는 메서드 호출을 제거하였습니다.

```java
    public List<Attend> getAttends(List<Integer> dayOfWeek) {
        return dayOfWeek.stream()
                .map(this::findByDay)
                .toList();
    }
```

sinb57 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,22 @@
+
+    public static boolean isHoliday(LocalDate targetDate) {
+        return Arrays.stream(Holiday.values())
+                .map(holiday -> LocalDate.of(Current.YEAR, holiday.month, holiday.date))
```
메서드를 호출할 때마다 LocalDate 객체가 생성될텐데, 재사용은 어려울까요?

jhan0121 replied:
Holiday에 date 멤버 변수를 추가하였고 `createDate`메서드를 활용하여 생성자에서 활용하도록 구성하였습니다. `isHoliday`를 호출할 때마다 생성하지 않고 해당 변수를 활용하도록 로직을 재구성하였습니다. 

```java
public enum Holiday {
    CHRISTMAS(12, 25);

    private final int month;
    private final int day;
    private final LocalDate date;

    Holiday(int month, int day) {
        this.month = month;
        this.day = day;
        this.date = createDate();
    }

    private LocalDate createDate() {
        return LocalDate.of(Current.YEAR, this.month, this.day);
    }

    public static boolean isHoliday(LocalDate targetDate) {
        return Arrays.stream(Holiday.values())
                .anyMatch(holiday -> holiday.isEqualDate(targetDate));
    }

    private boolean isEqualDate(LocalDate localDate) {
        return this.date.isEqual(localDate);
    }
}
```

sinb57 commented on src/main/java/view/DateTimeFormat.java:
```diff
@@ -0,0 +1,18 @@
+import java.time.format.DateTimeFormatter;
+
+public enum DateTimeFormat {
+    DATE(DateTimeFormatter.ofPattern("MM월 dd일 E요일")),
+    TIME(DateTimeFormatter.ofPattern("HH:mm"));
```
```suggestion
    DATE("MM월 dd일 E요일"),
    TIME("HH:mm");
```
DateTimeFormatter 생성 호출이 반복되네요. 중복 코드는 줄이기 위해 위와 같이 수정하는 것은 어려울까요?

jhan0121 replied:
멤버 변수로 String인 `dateTimePattern`과 DateTimeFormatter인 `dateTimeFormatter`로 재구성 하여 중복 코드를 줄이고 반복적으로 DateTimeFormatter가 생성되는 것을 방지하였습니다. 제안하신 방법을 적용하여 중복된 코드를 제거하였습니다!

```java
public enum DateTimeFormat {
    DATE("MM월 dd일 E요일"),
    TIME("HH:mm");

    private final String dateTimePattern;
    private final DateTimeFormatter dateTimeFormatter;

    DateTimeFormat(String dateTimePattern) {
        this.dateTimePattern = dateTimePattern;
        this.dateTimeFormatter = createDateTimeFormatter();
    }

    public DateTimeFormatter createDateTimeFormatter() {
        return DateTimeFormatter.ofPattern(dateTimePattern);
    }

    public DateTimeFormatter getDateTimeFormatter() {
        return dateTimeFormatter;
    }
}
```

sinb57 commented on src/main/java/domain/Current.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.List;
+import java.util.stream.IntStream;
+
+public enum Current {
+    TODAY(13);
+
+    public static final int YEAR = 2024;
+    public static final int MONTH = 12;
```
> 2024년 12월 한 달 동안 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다

요구사항을 보면, 현재는 12월 한정 시범 운영이지만 시스템이 확장될 가능성을 엿볼 수 있어요. 이 경우 현재의 구현에서 확장에 대응하려면 어떠한 변경이 있을지 생각해보시겠어요? 과연 현재의 코드는 변경에 열려 있나요?

jhan0121 replied:
현재 상황에서 12월 외에 다른 달로 확장해서 적용한다고 가정한다면 

3가지의 변화가 있을 것 같습니다.

### Current 변화
- Current에서 LocalDate.now() 적용해서 동적으로 오늘 날짜를 가져와야 함 -> 생성자에서 LocalDate 멤버 변수를 가지도록 변경
- static 상수인 `YEAR`와 `MONTH`가 동적으로 변화가 있어야 함 -> getter를 통해서 값을 전달하도록 변경

### Holiday 변화
- 12월 이외에 존재하는 공휴일 추가 -> enum 값 추가

`Current`의 경우에는 다른 도메인에 영향과 Current의 변경되는 코드의 양을 줄이기 위해서 아래와 같이 재구성하였습니다. 만약, 모든 달을 적용하게 되어서 `LocalDate.now()`를 사용해야 하는 경우에는 생성자에 있는 `LocalDate.of(2024, 12, 13)`부분을 대체하기만 하면 다른 코드에 영향 없이 작동되도록 하였습니다. 이와 같이 구성한다면 확장에 대응할 수 있을 것으로 생각합니다!

```java
public enum Current {
    TODAY;

    private final LocalDate date;

    Current() {
        this.date = LocalDate.of(2024, 12, 13);
    }

    public static boolean isDayOff(int day) {
        LocalDate targetDate = LocalDate.of(TODAY.getYear(), TODAY.getMonth(), day);
        return targetDate.getDayOfWeek().getValue() >= DayOfWeek.SATURDAY.getValue()
                || Holiday.isHoliday(targetDate);
    }

    public List<Integer> getAttendUntilDay() {
        return IntStream.range(1, getDay())
                .filter(day -> !isDayOff(day))
                .boxed()
                .toList();
    }

    public int getLengthOfMonth() {
        return getDate().lengthOfMonth();
    }

    public LocalDate getDate() {
        return date;
    }

    public int getYear() {
        return this.date.getYear();
    }

    public int getMonth() {
        return this.date.getMonthValue();
    }

    public int getDay() {
        return this.date.getDayOfMonth();
    }
}
```

sinb57 reviewed (changes requested):
안녕하세요 플린트~ 리뷰 반영하시느라 고생 많으셨어요!
앞서 남겼던 피드백과 더불어 추가 코멘트도 확인 부탁드려요~

jhan0121 committed:
refactor: AttendsTest 내 List<Attend>를 생성하는 메서드 추가

jhan0121 committed:
refactor: Command 약어 및 매개변수명 수정

jhan0121 committed:
refactor: Attends getAttends 메소드 stream 구조 최적화

jhan0121 committed:
refactor: 불필요한 매직 넘버 상수 제거

jhan0121 committed:
refactor: DateTimeFormat 구조 변경

- 멤버 변수 구성 수정

jhan0121 committed:
refactor: Holiday 구조 수정

- LocalDate 멤버 변수 추가
- isHoliday 로직 수정

jhan0121 committed:
refactor: OutputView 메서드 크기 분리 및 순서 조정

- printAttendanceResult 메서드 요구사항 만족을 위한 메서드 분할 수행
- 연관 메서드 간 위치 조정

jhan0121 committed:
refactor: OutputView에서 AttendResultDto 적용

jhan0121 committed:
refactor: AttendResultDto dto 패키지로 이동

jhan0121 committed:
style: 연관 메서드 간 거리 조정

- Attend
- AttendanceBook
- AttendanceController
- Attends
- InputView

jhan0121 committed:
Current 구조 재구성

- 확장성을 고려하여 구조 재구성

jhan0121 commented:
안녕하세요 이프!
말씀해주신 피드백 반영해서 코드 수정했습니다!

주요 변경 사항으로 6가지가 있습니다.

- 약어 사용 변수명 변경
- 과도한 매직넘버 상수화 조정하기(로직의 변수화 적용)
- 연관성 높은 메서드 간 거리 조정하기
- 도메인의 dto 의존 문제 해결하기
- 확장성 고려해서 재구성하기(Current 재구성)
- 중복되는 코드 재구성

sinb57 commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public void registerName(String name) {
+        attendsPerCrew.putIfAbsent(name, new Attends(new ArrayList<>()));
```
`putIfAbsent`를 사용하셨군요~ 👍 
`putIfAbsent`와 `computeIfAbsent`는 어떤 차이가 있나요?

jhan0121 replied:
## putIfAbsent
`putIfAbsent()`는 key 값이 존재하면 key에 해당되는 value를 반환하고, key 값이 존재하지 않으면 key와 value를 Map에 저장한 다음 value를 반환하는 메서드 입니다.
해당 메서드는 2개의 매개변수가 필요한 메서드 입니다. 1번째는 대상 map에 저장할 key를 할당하고, 2번째 매개변수는 대상 map에서 key가 존재하지 않을 때, put할 value를 할당합니다.

예시로 두 코드의 실행 방식은 동치입니다.(Map<String, Integer> map을 활용하는 것으로 가정하겠습니다.)

### putIfAbsent()
```java
String key = "100";
int value = map.putIfAbsent(key, Integer.parseInt(key);
```

### 동치 코드
```java
int value = map.get("100");
if (value == null) {
    value = map.put("100", 100);
}
```

## computeIfAbsent
`computeIfAbsent()`는  key 값이 존재하면 key에 해당되는 value를 반환하고, key 값이 존재하지 않으면 대상 function을 실행하여 받은 값을 value로 map에 저장한 다음 value를 반환하는 메서드 입니다.

해당 메서드는 2개의 매개변수가 필요한 메서드입니다. 1번째는 대상 map에 저장할 key를 할당하고, 2번째 매개변수는 저장할 key가 map에 존재하지 않는 경우, 실행하기 위한 메서드가 할당됩니다.

예시로 두 코드의 실행 방식은 동치입니다.(Map<String, Integer> map을 활용하는 것으로 가정하겠습니다.)

### computeIfAbsent()
```java
map.computeIfAbsent("10", key -> Integer.parseInt(key))
```

### 동치 코드
```java
Function<String, Integer> function = key -> Integer.parseInt(key);
String key = "100";
if (map.get(key) == null) {
    int value = function.apply(key);
    if (value != null) {
        map.put(key, value);
    }
}
```

즉, 정리하면 `putIfAbsent`에서는 2번째 매개변수로 값 자체를 넣어주고, `computeIfAbsent`에서는 2번째 매개변수로 key를 입력으로 하고 value를 리턴 하는 함수를 넣어주는 차이점이 존재합니다.

제 코드에서는 복잡한 로직이 아닌 생성자를 호출하는 정도라 `putIfAbsent`로 충분하다고 판단하여 `putIfAbsent`를 적용했었습니다!

sinb57 replied:
> 정리하면 putIfAbsent에서는 2번째 매개변수로 값 자체를 넣어주고, computeIfAbsent에서는 2번째 매개변수로 key를 입력으로 하고 value를 리턴 하는 함수를 넣어주는 차이점이 존재합니다.

조사는 잘 해주셨지만, 현재 상황에서는 객체 생성 비용을 절감하는 측면에서 computeIfAbsent가 나은 선택지예요.

```java
// name에 해당하는 value가 있음에도, Attends 객체를 생성. -> 사용하지 않는 객체 생성 -> GC에 의해 수거
attendsPerCrew.putIfAbsent(name, new Attends(new ArrayList<>()));

// name에 해당하는 value가 없어야, Attends 객체를 생성.
attendsPerCrew. computeIfAbsent(name, (it) -> new Attends(new ArrayList<>()));
```

sinb57 commented on src/main/java/domain/Attends.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void removeContainedAttend(Attend target) {
+        attends = attends.stream()
+                .filter(before -> !before.isDayEqual(target))
+                .collect(Collectors.toList());
```
attends 변수를 재할당하지 말고, removeIf를 활용해보면 어떨까요?

jhan0121 replied:
`removeIf`를 적용해 코드 로직을 단순화하였습니다!

```java
private void removeContainedAttend(Attend target) {
        attends.removeIf(before -> before.isDayEqual(target));
}
```

sinb57 reviewed (changes requested):
안녕하세요 플린트~ 고생 많으셨어요!! 코멘트 몇개 남겼는데 답변 부탁드려요~
따로 이슈 없다면, 다음 리뷰 요청 때 머지하겠습니다!!

jhan0121 committed:
fix: AttendanceController csv 파일 경로 수정

jhan0121 committed:
refactor: Attends removeContainedAttend 로직 수정

jhan0121 commented:
안녕하세요 이프!
소중한 피드백 감사합니다!

이번 기회에 정말 많은 것을 배울 수 있는 계기가 되었습니다.

말씀하신 피드백 반영 완료했습니다!

- putIfAbsent와 computeIfAbsent 간의 차이점
- removeIf() 적용

sinb57 reviewed (approved):
안녕하세요 플린트~ 1단계 진행하느라 고생 많으셨어요!!
피드백이 많았는데도 꼼꼼히 반영해주셔서 감사해요.
2단계 진행을 위해 1단계 머지할게요~!!
